File: Algo-4/Lab1/BST.c
#include<stdio.h>
#include<stdlib.h>

typedef struct BST
{
    int data; 
    struct BST* left;
    struct BST* right;
}BST;

BST* NewNode(int data)
{
    BST* node=(BST*)malloc(sizeof(BST));
    node->data=data;
    node->left=NULL;
    node->right=NULL;

    return node;
}

BST* insert(BST* root, int key)
{
    if(root==NULL) return NewNode(key);

    if(key<root->data)
        root->left=insert(root->left,key);
    else if(key>root->data)
        root->right=insert(root->right,key);
    
    return root;
}

BST* search(BST* root, int key)
{
    if(root==NULL || root->data==key)
        return root;
    if(key<root->data) return search(root->left,key);
    if(key>root->data) return search(root->right,key);
    else return NULL;
}

void inorder(BST* root)
{
    if(root!=NULL)
    {
        inorder(root->left);
		printf("%d  ", root->data);
		inorder(root->right);
    }
}

void preorder(BST* root)
{
	if(root!=NULL)
	{
		printf("%d  ", root->data);
		preorder(root->left);
		preorder(root->right);
	}
}

void postorder(BST* root)
{
	if(root!=NULL)
	{
		postorder(root->left);
		postorder(root->right);
		printf("%d  ", root->data);
	}
}

int main()
{
    BST* root=NULL;
    root=insert(root,50);
    int arr[]={30,20,70,40,60,80};
    for(int i=0; i<6; i++)
    	insert(root,arr[i]);

    printf("Inorder: ");
    inorder(root);
    printf("\n");

    printf("Preorder: ");
    preorder(root);
    printf("\n");

    printf("Postorder: ");
    postorder(root);
    printf("\n");

    if(search(root,45)==NULL)
    {
        printf("Element %d not in given BST.\n",45);
        insert(root,45);
        printf("Element %d inserted into the BST.\n",45);
    }
    else printf("Element %d is in given BST.\n",45);

    printf("Inorder: ");
    inorder(root);
    printf("\n");
}

File: AI-4/L1/BST.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

def insert(node, key):
    if node is None:
        return Node(key)

    if key < node.key:
        node.left = insert(node.left, key)
    else:
        node.right = insert(node.right, key)

    return node

def inorder(root):
    if root is not None:
        inorder(root.left)
        print(str(root.key), end=' ')
        inorder(root.right)

def preorder(root):
    if root is not None:
        print(str(root.key), end=' ')
        preorder(root.left)
        preorder(root.right)

def postorder(root):
    if root is not None:
        postorder(root.left)
        postorder(root.right)
        print(str(root.key), end=' ')

tree = [15,50,10,22,35,70,4,12,18,24,31,44,66,90]

root = None
root = insert(root, 25)
for i in tree:
    root = insert(root, i)

print('Inorder: ', end='')
inorder(root)
print('\nPreorder: ', end='')
preorder(root)
print('\nPostorder: ', end='')
postorder(root)